### PR TITLE
feat(mlx): audio-input training for VLM SFT, behind a per-family gate

### PR DESCRIPTION
**Pull request by Lyxot**
Adds audio-input support to MLX VLM SFT, behind a per-family allow gate, and qualifies four families: Gemma 3n, Gemma 4, Phi-4-multimodal and MiniCPM-o. Before this, an audio column either crashed collation or trained against misaligned features; audio rows are now either collated correctly or refused with a message that names what is wrong.

## Stacked on #963

**This is stacked on #963 ("fix(mlx): VLM training correctness"), which is not merged yet.** GitHub's diff is against `main`, so it shows that PR's commits too. The reviewable change here is the 8 commits on top of #963's tip `40b3f5b8`:

```bash
git diff 40b3f5b8..HEAD          # 5 files, 3589 insertions, 62 deletions
```

| | vs #963 tip (this PR) | vs `main` (what GitHub shows) |
|---|---|---|
| files | 5 | 6 |
| insertions | **3589** | 4596 |
| deletions | 62 | 92 |

Please review against `40b3f5b8`. I will retarget this once #963 lands.

## What changed

- **Collation.** Audio clips are extracted from dataset rows, grouped the way each processor assigns them to rows, and passed to the processor with the truncation arguments it actually honours. Rows whose audio and text cannot be paired are refused rather than trained.
- **Alignment is verified, not assumed.** Families whose placeholder run carries an identifying token are checked by counting runs. Families that state spans instead (MiniCPM-o reports `audio_bounds`, and its run interior is `<unk>`, which a genuine unknown word is indistinguishable from) are checked against those spans: one span per clip, inside the row's attended content, and naming a real placeholder run rather than ordinary text.
- **Loss masking.** Audio placeholder positions and the delimiters around them are excluded from the loss. For a stated-span family this cannot be done by token id, so the spans themselves drive the mask; without it every audio position was a supervised target.
- **Model-side merge.** Gemma 4's merge walks a flattened feature block with a running placeholder count, so a padded row's tail is consumed by the next row. It is patched per row. Gemma 4's placeholder count is also derived from the installed feature extractor rather than `ceil(duration / 40 ms)`, which runs one over whenever a duration lands near a frame boundary — 3 of 10 real LibriSpeech utterances.
- **Audio towers stay frozen**, including the projection that follows them, so a full fine-tune cannot pull them into the optimizer.
- **The gate.** `_AUDIO_QUALIFIED_FAMILIES` admits a family only at the exact mlx-vlm version it was probed on; Gemma 4 additionally requires transformers 5.5, since older releases expand its audio tokens differently. An unqualified family is refused with the versions that were verified.

## Public API

`unsloth_zoo.mlx.utils` gains two public names, recorded here because a consumer branch adopts them:

- `audio_input_capability(model, processor, texts=None) -> AudioInputCapability` — whether a loaded checkpoint can take audio, answered from behaviour rather than configuration. Audio content has to change what the processor returns (the same tone twice must agree, a second tone must differ, both share duration and construction, every call gets a fresh buffer) and the model has to carry audio modules, found by walking the module tree for the name. Result carries `capable`, `reason`, `processor_ok`, `model_ok`. Never raises, so it cannot fail a model load; anything unevaluable answers negative. Candidate prompts are the caller's, via `texts` — which marker a processor needs is prompt-rendering knowledge and the built-in candidate carries none.
- `audio_extractor_sampling_rate(processor)` — the promoted former private helper.

Nothing in training consults it. An empty audio payload already fails the per-batch pairing check on the same batch in the same call, so that message now points upstream instead, naming a checkpoint without an audio feature extractor as one example rather than as a diagnosis the check cannot make.

Measured against every audio-capable checkpoint in the local cache — gemma-3n-E2B, gemma-4-E4B (through its transformers 5.5 routing), MiniCPM-o-4.5 and Phi-4-multimodal answer capable; Qwen3-0.6B, Qwen2.5-VL-3B and the Ferox Phi-4 export answer not-capable. Every verdict matches expectation.

## Validation

Each family: 30 steps of LoRA on real LibriSpeech utterances, gradient checkpointing and CCE enabled, `dataset_order="torch_randperm"`. All four run to completion with finite losses and gradients.

| family | transformers | LoRA | step 1 | step 30 | mean loss |
|---|---|---|---|---|---|
| gemma4 | 5.5.0 | 35 × `q_proj` | 3.1281 | 1.0487 | 2.5040 |
| gemma3n | 4.57.6 | 30 × `q_proj` | 4.2405 | 1.9014 | 3.4067 |
| phi4mm | 4.57.6 | 32 × `qkv_proj` | 11.2239 | 5.0805 | 7.8304 |
| minicpmo | 4.57.6 | 36 × `q_proj` | 4.9293 | 2.2900 | 3.7666 |

Phi-4-multimodal's absolute loss is high because chance is 12.21 nats for its ~200k vocabulary; 28 of its 30 steps are below that.

**Gemma 4 against CUDA.** Same checkpoint on both machines, verified by digest rather than downloaded twice, and the same fixture and row order. The CUDA side runs unsloth's own stack (`FastVisionModel` + `get_peft_model` + `use_gradient_checkpointing="unsloth"`) on an RTX 4070 SUPER, both sides in bf16, neither side's adapters pinned.

| | MLX | CUDA | |
|---|---|---|---|
| mean loss (30) | 2.4388 | 2.4394 | −0.02% |
| mean grad norm (30) | 2.1523 | 2.2211 | −3.10% |
| step-1 loss | 3.1366 | 3.0965 | +1.30% |

图片

Both sides produce identical sequence lengths and identical supervised-token counts step for step, so the losses are computed over the same positions.

**One open question, stated plainly.** Step 1 is the only directly comparable point — `lora_B` starts at zero in both implementations, so the adapters contribute nothing and it is a pure forward-pass comparison of the same weights on the same tokens. The two differ by 1.30% (0.0401 nats), which is larger than bf16 accumulation-order differences alone produce. Quantization is not the cause (running MLX 4-bit instead moves it to +1.02%, the wrong direction), nor are differing supervised positions. The training trajectories agree closely, so this does not block the feature, but it is not yet explained. A text-only run of the same comparison would localise it to the audio path or to Gemma 4's forward generally.

## Known limits

- Prompt/completion collation is refused for stated-span families: the combine moves tokens the spans name, and nothing here has verified a relocation.
- With no literal audio marker in the rendered text, MiniCPM-o's processor places the run after the last user tag rather than where the audio part sat in the message. Row ownership is correct; intra-turn placement is not addressed.
- Clips longer than the feature extractor's window are refused rather than silently truncated.

## Test coverage

Tests are in `tests/test_mlx_vlm_label_masks.py`. Every added test was mutation-checked: reverting or weakening the production change it covers makes exactly that test fail. Test share of this PR's diff is 46%, which is above the usual bar; the audio path fails silently rather than loudly when alignment is wrong, so the coverage is deliberate.

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/utils.py [hunk 1]
```diff
@@ -6510,6 +8153,239 @@ def _vlm_family_encode_key(key):
...
+    poor training subjects, and the model-side merge assumes their output is a
+    fixed function of the clip. LoRA runs freeze everything by default, but a
+    full fine-tune would otherwise pull these parameters into the optimizer.
+
+    Returns the names of the modules that were frozen.
+    """
+    frozen = []
+    for attr in _AUDIO_TOWER_ATTRS:
+        for owner in (model, getattr(model, "language_model", None)):
+            module = getattr(owner, attr, None) if owner is not None else None
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/utils.py:8568]: Not fixing this one: the premise does not hold for the checkpoint it names.

I measured `freeze_audio_modules` against the real cached `microsoft/Phi-4-multimodal-instruct`, loaded the way this backend loads it:

```
freeze_audio_modules returned: ['audio_encoder', 'audio_projection']
audio-named modules: 788 | under the two frozen roots: 788 | outside: 0
audio params still trainable after freeze: 0
```

mlx-vlm exposes Phi-4's audio encoder and projection at the **top level** of the model, not under `embed_tokens_extend.audio_embed`. Both are in `_AUDIO_TOWER_ATTRS`, both are found, and `module.freeze(recurse=True)` covers everything beneath them — all 788 audio-named modules, leaving zero audio parameters in the optimizer.

The nested path in the comment appears to come from a test in this branch, whose docstring wrongly described that layout as Phi-4's. That string is a fixture exercising the name walk in `_model_carries_audio_modules` against a layout a fixed attribute list would miss; it was never a claim about the real family. I have corrected the docstring so it stops reading as one.

Residual, stated honestly and deliberately left out of scope: `freeze_audio_modules` does resolve by attribute root, so a *future* family that genuinely nests its tower would need the name walk here too. No qualified family does today, and widening it now would expand this PR past what it is for.

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/utils.py [hunk 2]
```diff
@@ -2137,6 +2342,27 @@ def _config_to_mapping(config):
...
+    restating them here.
+    """
+    image_id = _config_get(config, "image_token_index")
+    audio_id = _config_get(config, "audio_token_index")
+    if image_id is None or audio_id is None:
+        from mlx_vlm.models.phi4mm.config import ModelConfig as _Phi4MMConfig
+
+        fields = _Phi4MMConfig.__dataclass_fields__
+        if image_id is None:
+            image_id = fields["image_token_index"].default
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/utils.py:2414]: Not fixing this one. The premise is true but the consequence does not occur, and I verified that at the labels rather than at the resolver.

The recovered `image_token_index` is `-200`, a sentinel. Phi-4 compile preparation already resolves it, expands the image run, and `_expand_token_runs` writes `-100` for every replacement label -- so with the **pre-fix** ignore set all expanded `-200` positions end at `-100` and none is supervised. Threading the recovered id into the masking set therefore masks nothing.

Two further measurements on the shipped paths:

- with the default processor (`trust_remote_code=False`, mlx-vlm's vendored `Phi4MMProcessor`) the Phi-4 image expansion never runs for a collated row, and no negative id reaches `input_ids` or `labels`;
- with `trust_remote_code=True` the remote processor's image rows fail to collate outright (`Number of image tokens in prompt_token_ids (0) does not match number of images (1)`), so nothing reaches the labels there either.

I did implement this and then reverted it, because a change that fixes nothing still encodes a belief about the system that is not true. Recording the second point as a genuine finding for follow-up: Phi-4 image rows do not collate under `trust_remote_code=True`. That is unrelated to audio and out of scope here.

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/utils.py:8568]: Checked this independently and I agree with the author, with a stronger reason than the Phi-4 measurement alone gives.

The mechanism you describe is real: `freeze_audio_modules` walks only two owners, `model` and `model.language_model`, against five attribute names, so a tower one level deeper is missed. Constructed proof:

```
nested (embed_tokens_extend.audio_embed): frozen=[]                trainable audio params: 2
top-level (audio_encoder):                frozen=['audio_encoder'] trainable audio params: 0
```

But it is unreachable for every family the gate admits, and not by luck. In mlx-vlm 0.6.8 all four attach directly to the top-level `class Model`:

| family | line | attributes |
|---|---|---|
| gemma3n | `gemma3n.py:109-110` | `audio_tower`, `embed_audio` |
| gemma4 | `gemma4.py:57-61` | `audio_tower`, `embed_audio` |
| minicpmo | `minicpmo.py:228-230` | `audio_tower`, `audio_projection_layer` |
| phi4mm | `phi4mm.py:43-44` | `audio_encoder`, `audio_projection` |

All five names are in `_AUDIO_TOWER_ATTRS`. The `embed_tokens_extend.audio_embed` path you name is the Hugging Face layout, and mlx-vlm deliberately rewrites it away in `sanitize()` (`phi4mm.py:307`, comment `# Remap audio encoder weights (directly on Model now)`; minicpmo does the same at `:518` with `apm.` to `audio_tower.`). So the flattening is a stated contract of the loader, not an accident of one checkpoint, which is why the author's Phi-4 result generalises.

Marking this down as unreachable rather than wrong-in-principle. If a future qualified family keeps HF's nesting, the two-level walk would need to become a tree walk.

File: unsloth_zoo/mlx/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/utils.py:2414]: Agreeing with the author here too, verified at the labels rather than at the resolver.

Your premise holds: the recovered `image_token_index` is `-200` and it is not in the ignore set. But `_expand_token_runs` writes the ignore index over every replaced position unconditionally (`utils.py:2540-2541`):

```python
            new_ids.extend([int(token_id)] * repeat)
            ...
            if new_labels is not None:
                new_labels.extend([-100] * repeat)
```

so the expanded run is already unsupervised before any ignore-id set is consulted. Measured on a phi4mm batch:

```
ids:    [10, 11, -200, -200, -200, 12, 200011, 200011, 200011, 200011, 13]
labels: [10, 11, -100, -100, -100, 12,   -100,   -100,   -100,   -100, 13]
unmasked -200 label positions: []
```

I chased the one path that runs after compile preparation and could plausibly rebuild labels from the expanded ids, `train_on_responses_only`. It does not rebuild them: `_apply_response_mask_to_vlm_batch` always passes `mask_batch["labels"]`, so `use_old_labels` is true and `dataset_utils.py:498` copies the incoming labels rather than the ids. End to end with an entire row treated as one supervised span, the `-200` positions still come out `-100`.

A label carrier is also always present, so the `labels=False, raw=False` branch is unreachable: `_finalize_vlm_batch` assigns one on all three exits, and `_to_mx_vlm_batch` installs `_RAW_INPUT_IDS_FOR_LABELS` when `labels` is absent.

Threading the recovered id into the masking set would therefore mask nothing that is not already masked.

**Comment by danielhanchen** [comment]
Pushed `16fbaa94`: one regression this review turned up that is not in any of the inline threads.

An **audio-free** batch on an omni checkpoint was being refused. Both `_assert_audio_runs_intact_ids` and `_assert_audio_bounds_intact` cap a row's attended width at `max_seq_length`, but unlike the two clip-count checks beside them, neither cap was gated on the row actually carrying a clip. The soft-token branch is entered whenever the processor resolves an audio placeholder id, which gemma3n, gemma4, phi4mm and minicpmo all do whether or not the run uses audio.

Measured through the real collate, no audio anywhere in the data:

```
                                 merge-base   this branch   with the fix
width=40  max_seq_length=64          OK            OK            OK
width=40  max_seq_length=32          OK        ValueError        OK
width=120 max_seq_length=64          OK        ValueError        OK
```

The message a user got was about an audio placeholder run being cut, on a row with no audio in it. A rendered width above the cap is normal rather than an error, which `_plan_single_process_vlm_shapes` plans for explicitly, so this refused ordinary text and image training on all four qualified families. The merge-base column shows it is specific to this PR; #963 is unaffected.

Both caps now require `expected_clips`, which is what they were written for: they exist so a placeholder run is not cut unnoticed, and a row with no audio has no run to cut. The audio-bearing `over_cap` case in `test_unusable_audio_rows_are_rejected` is untouched and still passes. New test fails without the change, passes with it; suite otherwise unchanged at 259 passed.

Two things from the rest of the pass, neither of which needs action from you here.

**The suite does not run anywhere.** `mlx-ci.yml` is macOS-only, opt-in by the `mlx` label, and does not list `test_mlx_vlm_label_masks.py`, `test_mlx_shape_guard.py` or `test_mlx_batching_and_decay.py`. On Linux CI they skip at module level for want of a real MLX runtime. So the 2466 lines of tests here are currently executed by nothing. Worth adding the three files to that workflow.

They do run on Linux, incidentally: `mlx` publishes manylinux CPU wheels and `mlx-lm`/`mlx-vlm` are pure Python, so `mlx[cpu]` plus torch gives 259 of 260 passing on an x86_64 box. I also ran them on a real macos-14 runner and got the identical count, same single failure.

**That single failure is environment-dependent.** `test_qualified_families_carry_their_probed_requirements` asserts gemma4 is refused with `match="transformers"`, but on any machine with a current mlx-vlm the gate refuses earlier with the mlx-vlm message and never reaches the transformers floor. It assumes mlx-vlm exactly 0.4.4 and transformers below 5.5, neither declared, unlike `_qualify()` just above it which self-adapts. Reproduced on three environments with mlx-vlm 0.6.4, 0.6.7 and 0.6.8.

That is the canary for a bigger question about `_AUDIO_QUALIFIED_FAMILIES`: it admits a family only at exact string equality with `"0.4.4"`, while this package requires `mlx-vlm>=0.4.4`, Studio installs unpinned latest, and `mlx_repair.py` actively upgrades an old mlx-vlm. On a clean install today every audio row raises `NotImplementedError`. That is the gate doing what it says, so it is a release-sequencing question rather than a defect, but it is worth deciding before this merges.

**Comment by danielhanchen** [comment]
#963 landed as a squash (`bee10411`), so git could no longer match this branch's copies of those commits to it and the PR went to CONFLICTING on `unsloth_zoo/mlx/utils.py` and `tests/test_mlx_vlm_label_masks.py`. Merged main in at `0eb946b1` to clear that, no rebase or force push, so the branch history is intact.

Every conflict resolved to the branch side, and that is provable rather than a judgement call: the squash is byte-identical to #963's approved head `40b3f5b8` for all three mlx files, so the branch already carried main's content there plus this PR's work on top. After resolution both files match `33a23a9a` exactly. Main's other changes in the window (`saving_utils`, `rl_replacements`, `diffusion_studio`) touch no mlx file and merged cleanly.

Re-verified on the merged tree: 275 passed, 1 failed, the failure being the mlx-vlm version gate discussed above rather than anything from the merge. Simulation suite 145 passing.

One thing worth flagging now that #963 is in: the shared K/V truncation fix (`6afaadac`) is on this branch, not on #963, so `main` currently builds legacy cache slots for a natively shared Gemma4 backbone and `zip(layers, cache)` drops the shared tail of the stack. Reproduced against the merge commit itself:

```
main bee10411   native shared kv: True   layers/slots: 4/2   TRUNCATES
this branch     native shared kv: True   layers/slots: 4/0   fixed
```

So this PR is what closes it.

**Comment by danielhanchen** [comment]
@codex review
